File: Assets/Bundles/Shared/Resources/TS/Network/NetworkAPI.lua
-- Compiled with unity-ts v2.1.0-75
local RunUtil = require("Shared/TS/Util/RunUtil").RunUtil
local NetworkChannel
do
	local _inverse = {}
	NetworkChannel = setmetatable({}, {
		__index = _inverse,
	})
	NetworkChannel.Reliable = 0
	_inverse[0] = "Reliable"
	NetworkChannel.Unreliable = 1
	_inverse[1] = "Unreliable"
end
local MAX_QUEUE = 10000
local callbacksById = {}
local queuedDataById = {}
local function addToQueue(msg)
	local id = tostring(msg.i)
	local queue = queuedDataById[id]
	if queue == nil then
		queue = {}
		local _queue = queue
		queuedDataById[id] = _queue
	end
	if #queue >= MAX_QUEUE then
		print("REMOTE EVENT: MAX REQUEST QUEUE REACHED; DROPPING FUTURE MESSAGES")
		return nil
	end
	local _queue = queue
	local _msg = msg
	table.insert(_queue, _msg)
end
local function InitNet()
	if RunUtil:IsServer() then
		-- Server
		NetworkCore.Net:OnBroadcastFromClientAction(function(clientId, blob)
			local msg = blob:Decode()
			local id = msg.i
			local data = msg.d
			local _arg0 = tostring(id)
			local callbacks = callbacksById[_arg0]
			if callbacks == nil then
				return nil
			end
			for _, callback in callbacks do
				callback.callback(clientId, unpack(data))
			end
		end)
	else
		-- Client
		NetworkCore.Net:OnBroadcastFromServerAction(function(blob)
			local msg = blob:Decode()
			local id = msg.i
			local data = msg.d
			local _arg0 = tostring(id)
			local callbacks = callbacksById[_arg0]
			if callbacks == nil then
				-- Queue data
				addToQueue(msg)
				return nil
			end
			for _, callback in callbacks do
				callback.callback(unpack(data))
			end
		end)
	end
end
local function pack(id, args)
	return BinaryBlob.new({
		i = id,
		d = args,
	})
end
local function fireServer(id, args, channel)
	local msg = pack(id, args)
	NetworkCore.Net:BroadcastToServer(msg, if channel == NetworkChannel.Reliable then 1 else 0)
end
local function fireAllClients(id, args, channel)
	local msg = pack(id, args)
	NetworkCore.Net:BroadcastToAllClients(msg, if channel == NetworkChannel.Reliable then 1 else 0)
end
local function fireClient(id, clientId, args, channel)
	local msg = pack(id, args)
	NetworkCore.Net:BroadcastToClient(clientId, msg, if channel == NetworkChannel.Reliable then 1 else 0)
end
local function fireExcept(id, ignoredClientId, args, channel)
	local msg = pack(id, args)
	NetworkCore.Net:BroadcastToAllExceptClient(ignoredClientId, msg, if channel == NetworkChannel.Reliable then 1 else 0)
end
local function fireClients(id, clientIds, args, channel)
	local msg = pack(id, args)
	NetworkCore.Net:BroadcastToClients(clientIds, msg, if channel == NetworkChannel.Reliable then 1 else 0)
end
local function connect(id, callback)
	local connected = true
	local _arg0 = tostring(id)
	local callbacks = callbacksById[_arg0]
	if callbacks == nil then
		callbacks = {}
		local _arg0_1 = tostring(id)
		local _callbacks = callbacks
		callbacksById[_arg0_1] = _callbacks
	end
	-- Wrap callback in a unique object:
	local callbackItem = {
		callback = callback,
	}
	table.insert(callbacks, callbackItem)
	-- Invoke callback with any queued data:
	local _arg0_1 = tostring(id)
	local queue = queuedDataById[_arg0_1]
	if queue ~= nil then
		for _, msg in queue do
			callback(unpack(msg.d))
		end
		local _arg0_2 = tostring(id)
		queuedDataById[_arg0_2] = nil
	end
	-- Disconnect function
	return function()
		if not connected then
			return nil
		end
		connected = false
		local _arg0_2 = tostring(id)
		local callbacksDisconnect = callbacksById[_arg0_2]
		if callbacksDisconnect == nil then
			return nil
		end
		local index = (table.find(callbacksDisconnect, callbackItem) or 0) - 1
		if index == -1 then
			return nil
		end
		table.remove(callbacksDisconnect, index + 1)
	end
end
local NetworkAPI = {
	fireServer = fireServer,
	fireAllClients = fireAllClients,
	fireClient = fireClient,
	fireClients = fireClients,
	fireExcept = fireExcept,
	connect = connect,
}
-- initNet();
local default = NetworkAPI
return {
	InitNet = InitNet,
	NetworkChannel = NetworkChannel,
	default = default,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/@Easy/Core/Core~/src/Shared/Network/NetworkAPI.ts
import { Airship } from "Shared/Airship";
import { Player } from "Shared/Player/Player";
import { RunUtil } from "Shared/Util/RunUtil";

export enum NetworkChannel {
	Reliable,
	Unreliable,
}

interface BlobData {
	/** ID. */
	i: number;

	/** Data. */
	d: unknown[];
}

interface CallbackItem {
	callback: Callback;
}

const MAX_QUEUE = 10000;

const callbacksByIdServer = new Map<string, Array<CallbackItem>>();
const callbacksByIdClient = new Map<string, Array<CallbackItem>>();
const queuedDataById = new Map<string, Array<BlobData>>();

function addToQueue(msg: BlobData) {
	const id = tostring(msg.i);
	let queue = queuedDataById.get(id);
	if (queue === undefined) {
		queue = [];
		queuedDataById.set(id, queue);
	}
	if (queue.size() >= MAX_QUEUE) {
		warn("REMOTE EVENT: MAX REQUEST QUEUE REACHED; DROPPING FUTURE MESSAGES");
		return;
	}
	queue.push(msg);
}

export function InitNet() {
	if (RunUtil.IsServer()) {
		NetworkCore.Net.OnBroadcastFromClientAction((clientId, blob) => {
			const msg = blob.Decode() as BlobData;
			const id = msg.i;
			const data = msg.d;
			const callbacks = callbacksByIdServer.get(tostring(id));
			if (callbacks === undefined) return;
			const waitForClient = Airship.players.WaitForClientIdIncludePending(clientId);
			waitForClient.then((player) => {
				if (!player) {
					warn(debug.traceback(`Dropping request from client ${clientId}: Failed to find Player object`));
					return;
				}
				for (const callback of callbacks) {
					task.spawn(callback.callback, player, ...data);
				}
			});
		});
	}
	if (RunUtil.IsClient()) {
		NetworkCore.Net.OnBroadcastFromServerAction((blob) => {
			const msg = blob.Decode() as BlobData;
			const id = msg.i;
			const data = msg.d;
			const callbacks = callbacksByIdClient.get(tostring(id));
			if (callbacks === undefined) {
				// Queue data
				addToQueue(msg);
				return;
			}
			for (const callback of callbacks) {
				task.spawn(callback.callback, ...data);
			}
		});
	}
}

function pack(id: number, args: unknown[]) {
	return new BinaryBlob({ i: id, d: args } satisfies BlobData);
}

function fireServer(id: number, args: unknown[], channel: NetworkChannel) {
	const msg = pack(id, args);
	NetworkCore.Net.BroadcastToServer(msg, channel === NetworkChannel.Reliable ? 1 : 0);
}

function fireAllClients(id: number, args: unknown[], channel: NetworkChannel) {
	const msg = pack(id, args);
	NetworkCore.Net.BroadcastToAllClients(msg, channel === NetworkChannel.Reliable ? 1 : 0);
}

function fireClient(id: number, player: Player, args: unknown[], channel: NetworkChannel) {
	const msg = pack(id, args);
	NetworkCore.Net.BroadcastToClient(player.clientId, msg, channel === NetworkChannel.Reliable ? 1 : 0);
}

function fireExcept(id: number, ignorePlayer: Player, args: unknown[], channel: NetworkChannel) {
	const msg = pack(id, args);
	NetworkCore.Net.BroadcastToAllExceptClient(ignorePlayer.clientId, msg, channel === NetworkChannel.Reliable ? 1 : 0);
}

function fireClients(id: number, players: Player[], args: unknown[], channel: NetworkChannel) {
	const msg = pack(id, args);
	const clientIds = players.map((player) => player.clientId);
	NetworkCore.Net.BroadcastToClients(
		clientIds as unknown as CSArray<number>,
		msg,
		channel === NetworkChannel.Reliable ? 1 : 0,
	);
}

function connect(asServer: boolean, id: number, callback: Callback): () => void {
	const callbacksById = asServer ? callbacksByIdServer : callbacksByIdClient;

	let connected = true;
	let callbacks = callbacksById.get(tostring(id));
	if (callbacks === undefined) {
		callbacks = [];
		callbacksById.set(tostring(id), callbacks);
	}

	// Wrap callback in a unique object:
	const callbackItem: CallbackItem = { callback };
	callbacks.push(callbackItem);

	// Invoke callback with any queued data:
	const queue = queuedDataById.get(tostring(id));
	if (queue !== undefined) {
		for (const msg of queue) {
			callback(...msg.d);
		}
		queuedDataById.delete(tostring(id));
	}

	// Disconnect function
	return () => {
		if (!connected) return;
		connected = false;
		const callbacksDisconnect = callbacksById.get(tostring(id));
		if (callbacksDisconnect === undefined) return;
		const index = callbacksDisconnect.indexOf(callbackItem);
		if (index === -1) return;
		callbacksDisconnect.remove(index);
	};
}

const NetworkAPI = {
	fireServer,
	fireAllClients,
	fireClient,
	fireClients,
	fireExcept,
	connect,
};

export default NetworkAPI;
